### DISCUSSION TITLE
is langium can validate  through different file by default

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-langium/langium#1461] [Q&A]
is langium can validate through different file by default,i define a function named CalculateSum578 at aa.st,refer it at struct.st,i change name to CalculateSum or CalculateSum5 or CalculateSum57 or CalculateSum578,all of these seems did not report error,
but change to other name seems to report my custom validation error,a little confused
![image](https://github.com/eclipse-langium/langium/assets/26995040/ed95a1cb-a90f-43f6-a2f3-bac6fa9415fe)
![image](https://github.com/eclipse-langium/langium/assets/26995040/f095d1fb-0ae7-4f49-918b-292e266638f1)

**Comment by zrwsmd** [comment]
also,if i change function name CalculateSum578 to CalculateSum666,old name CalculateSum578 seems to still stored in my local map,how to remove it, i guess it related to langium lifecycle

**Reply by cdietrich** [reply]
you might want to look into `WorkspaceCache` and. `DocumentCache`
as replacements for these map

**Reply by zrwsmd** [reply]
what is WorkspaceCache and DocumentCache

**Reply by cdietrich** [reply]
the are cache classes in langium that get invalidated on document or workspace changes

**Reply by zrwsmd** [reply]
so how to use this cache in my custom validation method 'checkElement',
![image](https://github.com/eclipse-langium/langium/assets/26995040/5d0bb4c8-6675-45f1-a846-31ac2cfde4ff)


export function registerValidationChecks(services: StatemachineServices) {
    const cache = new WorkspaceCache<string, string>(services.shared);
    const registry = services.validation.ValidationRegistry;
    const validator = services.validation.StValidator;
    const checks: ValidationChecks<StAstType> = {
        St: validator.checkElement
    };
    registry.register(checks, validator);
}

export class StValidator {
    checkElement(st: St, accept: ValidationAcceptor): void {

}

**Reply by cdietrich** [reply]
simply check the other usages in the langium codebase

**Comment by msujew** [comment]
Can you share the validation code and maybe your grammar as well? By default, the validation should be retriggert on files that have references to changed files (or have been changed themselves).

**Reply by zrwsmd** [reply]
validation code is large ,1700 lines
assuming i have two st file in my workspace,aa.st and bb.st
aa.st define a functionBlock named TemperatureConverter,it have method ConvertCelsiusToFahrenheit
![image](https://github.com/eclipse-langium/langium/assets/26995040/f2f2c3db-8d07-4c7c-be43-f8a403395912)

bb.st refer this functionBlock 
![image](https://github.com/eclipse-langium/langium/assets/26995040/932fa3c6-0850-4ab8-b2ad-a45d442e6667)

it validate well.

but when i change aa.st functionBlock name 'TemperatureConverter' to 'TemperatureConverter55',bb.st seems did not report error,how should i retrigger it

**Reply by msujew** [reply]
Is there a real grammar cross reference between those elements? That's why I'm asking about the grammar.

**Reply by zrwsmd** [reply]
no

**Reply by msujew** [reply]
Well, how else is Langium supposed to identify whether to revalidate the file? You need to add a cross reference there. Langium won't just revalidate all files if one has changed, as that would lead to major performance issues.

**Reply by zrwsmd** [reply]
if i do not want to use cross reference,since a parse rule can be used by other parse rule,which is not easy to change.
if i define a  map at a validation file  which store my parse rule info(e.g st function block info),how can this map store all .st file  function block info so that i can define my custom validation code .
at moment, if i define a function block at aa.st,but refer it at bb.st, i can not get function block info,i can only get it at aa.st file,how can this map is shared by all st files

**Reply by msujew** [reply]
You can adjust the [`isAffected`](https://github.com/eclipse-langium/langium/blob/26cca7ae2448e5b579244e76e2aeb11bd8550927/packages/langium/src/workspace/index-manager.ts#L168-L174) computation of the `IndexManager`. It will be used when calculating which files need to be revalidated when another file has changed.

**Reply by zrwsmd** [reply]
i want to use my custom validation code instead of cross reference,i define a map at validation code
![image](https://github.com/eclipse-langium/langium/assets/26995040/d3c53c3f-99e1-4317-b169-57b99462f544)


 which key is function block name(e.g 'TemperatureConverter'),value is function block itself,i test like below screenshot
![image](https://github.com/eclipse-langium/langium/assets/26995040/edcabfd4-d8a5-4aba-9002-f54e3299765f)


i have three st files,aa.st ,bb.st and cc.st, i open this three files by order aa.st,bb.st ,cc.st ,like below screenshot,aa.st now can only get 
one function block info,
bb.st can get two function block info,
cc.st can get three function block info
![image](https://github.com/eclipse-langium/langium/assets/26995040/ea0298f4-959c-48b3-beb7-cecdcb6a0fd9)
![image](https://github.com/eclipse-langium/langium/assets/26995040/2988b92a-f175-4911-8cd2-b64d73c3c627)
![image](https://github.com/eclipse-langium/langium/assets/26995040/ee49ceb0-cd8d-4de9-b8fb-811ef88999fa)
how can each file get this threee function block info
